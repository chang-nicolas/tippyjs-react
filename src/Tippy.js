import React, {cloneElement, useState} from 'react';
import PropTypes from 'prop-types';
import {createPortal} from 'react-dom';
import {
  preserveRef,
  ssrSafeCreateDiv,
  toDataAttributes,
  updateClassName,
  deepPreserveProps,
} from './utils';
import {useInstance, useIsomorphicLayoutEffect} from './util-hooks';

export default function TippyGenerator(tippy) {
  function Tippy({
    children,
    content,
    className,
    visible,
    singleton,
    render,
    disabled = false,
    ignoreAttributes = true,
    // Filter React development reserved props
    // added by babel-preset-react dev plugins:
    // transform-react-jsx-self and transform-react-jsx-source
    __source,
    __self,
    ...restOfNativeProps
  }) {
    const isControlledMode = visible !== undefined;
    const isSingletonMode = singleton !== undefined;

    const [attrs, setAttrs] = useState({});
    const [mounted, setMounted] = useState(false);
    const [singletonContent, setSingletonContent] = useState();
    const component = useInstance(() => ({
      container: ssrSafeCreateDiv(),
      renders: 1,
    }));

    const props = {
      ignoreAttributes,
      hideOnClick: !isControlledMode,
      ...restOfNativeProps,
      content: component.container,
    };

    if (isControlledMode) {
      props.trigger = 'manual';
    }

    if (isSingletonMode) {
      disabled = true;
    }

    let computedProps = props;
    const plugins = props.plugins || [];

    if (render) {
      computedProps = {
        ...props,
        plugins: isSingletonMode
          ? [
              ...plugins,
              {
                fn: () => ({
                  onTrigger(_, event) {
                    const {content} = singleton.data.children.find(
                      ({instance}) =>
                        instance.reference === event.currentTarget,
                    );
                    setSingletonContent(content);
                  },
                }),
              },
            ]
          : plugins,
        render: () => ({popper: component.container}),
      };
    }

    const deps = children ? [children.type] : [];

    // CREATE
    useIsomorphicLayoutEffect(() => {
      const instance = tippy(
        component.ref || ssrSafeCreateDiv(),
        computedProps,
      );

      component.instance = instance;

      if (disabled) {
        instance.disable();
      }

      if (visible) {
        instance.show();
      }

      if (isSingletonMode) {
        singleton.hook({
          instance,
          content,
          props: computedProps,
        });
      }

      setMounted(true);

      return () => {
        instance.destroy();
      };
    }, deps);

    // UPDATE
    useIsomorphicLayoutEffect(() => {
      // Prevent this effect from running on 1st render
      if (component.renders === 1) {
        component.renders++;
        return;
      }

      const {instance} = component;

      instance.setProps(deepPreserveProps(instance.props, props));

      if (disabled) {
        instance.disable();
      } else {
        instance.enable();
      }

      if (isControlledMode) {
        if (visible) {
          instance.show();
        } else {
          instance.hide();
        }
      }

      if (isSingletonMode) {
        singleton.hook({
          instance,
          content,
          props: computedProps,
        });
      }
    });

    useIsomorphicLayoutEffect(() => {
      if (!render) {
        return;
      }

      const {instance} = component;

      instance.setProps({
        popperOptions: {
          ...instance.props.popperOptions,
          modifiers: [
            ...(instance.props.popperOptions?.modifiers || []),
            {
              name: '$$tippyReact',
              enabled: true,
              phase: 'beforeWrite',
              requires: ['computeStyles'],
              fn({state}) {
                const hideData = state.modifiersData?.hide;

                // WARNING: this is a high-risk path that can cause an infinite
                // loop. This expression _must_ evaluate to false when required
                if (
                  attrs.placement !== state.placement ||
                  attrs.referenceHidden !== hideData?.isReferenceHidden ||
                  attrs.escaped !== hideData?.hasPopperEscaped
                ) {
                  setAttrs({
                    placement: state.placement,
                    referenceHidden: hideData?.isReferenceHidden,
                    escaped: hideData?.hasPopperEscaped,
                  });
                }

                state.attributes.popper = {};
              },
            },
          ],
        },
      });
    }, [attrs.placement, attrs.referenceHidden, attrs.escaped, ...deps]);

    useIsomorphicLayoutEffect(() => {
      if (className) {
        if (render) {
          if (process.env.NODE_ENV !== 'production') {
            console.warn(
              [
                '@tippyjs/react: Cannot use `className` prop in conjunction',
                'with the `render` prop. Place the className on the element',
                'you are rendering.',
              ].join(' '),
            );
          }

          return;
        }

        const box = component.instance.popper.firstElementChild;

        updateClassName(box, 'add', className);

        return () => {
          updateClassName(box, 'remove', className);
        };
      }
    }, [className, ...deps]);

    return (
      <>
        {children
          ? cloneElement(children, {
              ref(node) {
                component.ref = node;
                preserveRef(children.ref, node);
              },
            })
          : null}
        {mounted &&
          createPortal(
            render
              ? render(toDataAttributes(attrs), singletonContent)
              : content,
            component.container,
          )}
      </>
    );
  }

  if (process.env.NODE_ENV !== 'production') {
    const ContentType = PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
      PropTypes.element,
    ]);

    Tippy.propTypes = {
      children: PropTypes.element,
      content: PropTypes.oneOfType([
        ContentType,
        PropTypes.arrayOf(ContentType),
      ]),
      render: PropTypes.func,
      visible: PropTypes.bool,
      disabled: PropTypes.bool,
      className: PropTypes.string,
      singleton: PropTypes.object,
    };
  }

  return Tippy;
}
