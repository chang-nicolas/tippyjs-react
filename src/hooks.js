import {
  useInstance,
  useUpdateClassName,
  useIsomorphicLayoutEffect,
} from './util-hooks';
import {createSingleton} from 'tippy.js';

export function useSingleton({
  className,
  enabled = true,
  ignoreAttributes = true,
  ...restOfNativeProps
} = {}) {
  const component = useInstance({
    instance: null,
    instances: [],
    renders: 1,
  });

  const props = {
    ignoreAttributes,
    ...restOfNativeProps,
  };

  const deps = [component.instances.length];

  useIsomorphicLayoutEffect(() => {
    const {instances} = component;
    const instance = createSingleton(instances, props);

    component.instance = instance;

    if (!enabled) {
      instance.disable();
    }

    return () => {
      instance.destroy();
      component.instances = instances.filter(i => !i.state.isDestroyed);
    };
  }, deps);

  useIsomorphicLayoutEffect(() => {
    if (component.renders === 1) {
      component.renders++;
      return;
    }

    const {instance} = component;

    instance.setProps(props);

    if (enabled) {
      instance.enable();
    } else {
      instance.disable();
    }
  });

  useUpdateClassName(component, className, deps);

  return instance => {
    component.instances.push(instance);
  };
}
