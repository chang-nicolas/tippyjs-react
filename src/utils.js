export const isBrowser =
  typeof window !== 'undefined' && typeof document !== 'undefined';

export function preserveRef(ref, node) {
  if (ref) {
    if (typeof ref === 'function') {
      ref(node);
    }
    if ({}.hasOwnProperty.call(ref, 'current')) {
      ref.current = node;
    }
  }
}

export function ssrSafeCreateDiv() {
  return isBrowser && document.createElement('div');
}

export function updateClassName(tooltip, action, classNames) {
  classNames.split(/\s+/).forEach(name => {
    if (name) {
      tooltip.classList[action](name);
    }
  });
}

export function toDataAttributes(attrs) {
  const dataAttrs = {
    'data-placement': attrs.placement,
  };

  if (attrs.referenceHidden) {
    dataAttrs['data-reference-hidden'] = '';
  }

  if (attrs.escaped) {
    dataAttrs['data-escaped'] = '';
  }

  return dataAttrs;
}
