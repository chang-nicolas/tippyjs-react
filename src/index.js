import tippy from 'tippy.js';
import Tippy from './Tippy';
import {useSingleton} from './hooks';

export default Tippy;
export {useSingleton, tippy};
