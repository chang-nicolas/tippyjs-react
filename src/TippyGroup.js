import { Children, cloneElement, useEffect } from 'react'
import PropTypes from 'prop-types'
import tippy from 'tippy.js'
import { useInstance } from './utils'

export default function TippyGroup({ children, ...props }) {
  const component = useInstance({ instances: [] })

  useEffect(() => {
    component.instances = component.instances.filter(i => !i.state.isDestroyed)
    tippy.group(component.instances, props)
  })

  return Children.map(children, child => {
    return cloneElement(child, {
      onCreate(instance) {
        if (child.props.onCreate) {
          child.props.onCreate(instance)
        }

        component.instances.push(instance)
      },
    })
  })
}

if (process.env.NODE_ENV !== 'production') {
  TippyGroup.propTypes = {
    children: PropTypes.arrayOf(PropTypes.element).isRequired,
  }
}
