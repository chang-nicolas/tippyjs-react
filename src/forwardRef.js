import React, {cloneElement, forwardRef} from 'react';
import {preserveRef} from './utils';

export default Tippy =>
  forwardRef(function TippyWrapper({children = <div />, ...props}, ref) {
    return (
      <Tippy {...props}>
        {cloneElement(children, {
          ref(node) {
            preserveRef(ref, node);
            preserveRef(children.ref, node);
          },
        })}
      </Tippy>
    );
  });
