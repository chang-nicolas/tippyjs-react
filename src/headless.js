import React, {cloneElement, forwardRef} from 'react';
import tippy from 'tippy.js/headless';
import {useSingleton} from './hooks';
import TippyGenerator from './Tippy';
import {preserveRef} from './utils';

const Tippy = TippyGenerator(tippy);

export default forwardRef(function TippyWrapper({children, ...props}, ref) {
  return (
    <Tippy {...props}>
      {cloneElement(children, {
        ref(node) {
          preserveRef(ref, node);
          preserveRef(children.ref, node);
        },
      })}
    </Tippy>
  );
});

export {useSingleton, tippy};
