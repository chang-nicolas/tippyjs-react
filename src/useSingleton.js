import {useMutableBox, useIsomorphicLayoutEffect} from './util-hooks';
import {deepPreserveProps, updateClassName} from './utils';
import {useMemo} from 'react';

export default function useSingletonGenerator(createSingleton) {
  return function useSingleton({disabled = false, overrides = []} = {}) {
    const mutableBox = useMutableBox({
      children: [],
      renders: 1,
    });

    useIsomorphicLayoutEffect(() => {
      const {children, sourceData} = mutableBox;

      if (!sourceData) {
        if (process.env.NODE_ENV !== 'production') {
          console.error(
            [
              '@tippyjs/react: The `source` variable from `useSingleton()` has',
              'not been passed to a <Tippy /> component.',
            ].join(' '),
          );
        }

        return;
      }

      const instance = createSingleton(
        children.map(child => child.instance),
        {
          ...sourceData.props,
          popperOptions: sourceData.instance.props.popperOptions,
          overrides,
        },
      );

      mutableBox.instance = instance;

      if (disabled) {
        instance.disable();
      }

      return () => {
        instance.destroy();
        mutableBox.children = children.filter(
          ({instance}) => !instance.state.isDestroyed,
        );
      };
    }, []);

    useIsomorphicLayoutEffect(() => {
      if (mutableBox.renders === 1) {
        mutableBox.renders++;
        return;
      }

      const {children, instance, sourceData} = mutableBox;

      if (!(instance && sourceData)) {
        return;
      }

      const {content, ...props} = sourceData.props;
      instance.setProps(
        deepPreserveProps(instance, {
          ...props,
          overrides,
        }),
      );

      instance.setInstances(children.map(child => child.instance));

      if (disabled) {
        instance.disable();
      } else {
        instance.enable();
      }
    });

    useIsomorphicLayoutEffect(() => {
      const className = mutableBox.sourceData?.className;

      if (className) {
        if (mutableBox.sourceData?.props.render) {
          if (process.env.NODE_ENV !== 'production') {
            console.warn(
              [
                '@tippyjs/react: Cannot use `className` prop in conjunction',
                'with the `render` prop. Place the className on the element',
                'you are rendering.',
              ].join(' '),
            );
          }

          return;
        }

        const box = mutableBox.instance.popper.firstElementChild;

        updateClassName(box, 'add', className);

        return () => {
          updateClassName(box, 'remove', className);
        };
      }
    });

    return useMemo(() => {
      const source = {
        data: mutableBox,
        hook(data) {
          mutableBox.sourceData = data;
        },
        cleanup() {
          mutableBox.sourceData = null;
        },
      };

      const target = {
        hook(data) {
          if (
            !mutableBox.children.find(
              ({instance}) => data.instance === instance,
            )
          ) {
            mutableBox.children.push(data);
          }
        },
        cleanup(instance) {
          mutableBox.children = mutableBox.children.filter(
            data => data.instance !== instance,
          );
        },
      };

      return [source, target];
    }, []);
  };
}
