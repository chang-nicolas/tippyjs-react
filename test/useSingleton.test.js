/* eslint-disable */
import React from 'react';
import Tippy, {useSingleton} from '../src';
import {useSingleton as useSingletonHeadless} from '../src/headless';
import {render, cleanup, fireEvent} from '@testing-library/react';

jest.useFakeTimers();

afterEach(cleanup);

let instance;
function onCreate(i) {
  instance = i;
}

it('changes the singleton content correctly', () => {
  function App() {
    const [source, target] = useSingleton();

    return (
      <>
        <Tippy
          onCreate={onCreate}
          singleton={source}
          trigger="click"
          hideOnClick={false}
        />
        <Tippy content="a" singleton={target}>
          <button data-testid="a" />
        </Tippy>
        <Tippy content="b" singleton={target}>
          <button data-testid="b" />
        </Tippy>
      </>
    );
  }

  const {getByTestId} = render(<App />);

  const buttonA = getByTestId('a');
  const buttonB = getByTestId('b');

  fireEvent.click(buttonA);

  expect(instance.state.isVisible).toBe(true);
  expect(instance.props.content.textContent).toBe('a');

  fireEvent.click(buttonB);

  expect(instance.props.content.textContent).toBe('b');
});

describe('disabled prop', () => {
  function App({disabled}) {
    const [source, target] = useSingleton({disabled});

    return (
      <>
        <Tippy
          onCreate={onCreate}
          singleton={source}
          trigger="click"
          hideOnClick={false}
        />
        <Tippy content="a" singleton={target}>
          <button data-testid="a" />
        </Tippy>
        <Tippy content="b" singleton={target}>
          <button data-testid="b" />
        </Tippy>
      </>
    );
  }

  it('it is set correctly on mount and on updates', () => {
    const {getByTestId, rerender} = render(<App disabled={true} />);
    const buttonA = getByTestId('a');

    fireEvent.click(buttonA);

    expect(instance.state.isVisible).toBe(false);

    rerender(<App disabled={false} />);

    fireEvent.click(buttonA);

    expect(instance.state.isVisible).toBe(true);

    rerender(<App disabled={true} />);

    fireEvent.click(buttonA);

    expect(instance.state.isVisible).toBe(false);
  });
});

describe('useSingleton headless mode', () => {
  function App() {
    const [source, target] = useSingletonHeadless();

    return (
      <>
        <Tippy
          onCreate={onCreate}
          render={(attrs, content) => <div {...attrs}>{content}</div>}
          singleton={source}
          trigger="click"
          hideOnClick={false}
        />
        <Tippy content="a" singleton={target}>
          <button data-testid="a" />
        </Tippy>
        <Tippy content="b" singleton={target}>
          <button data-testid="b" />
        </Tippy>
      </>
    );
  }

  it('updates content correctly', () => {
    const {getByTestId} = render(<App />);

    const buttonA = getByTestId('a');
    const buttonB = getByTestId('b');

    fireEvent.click(buttonA);

    expect(instance.state.isVisible).toBe(true);
    expect(instance.popper).toMatchSnapshot();

    fireEvent.click(buttonB);

    expect(instance.popper).toMatchSnapshot();
  });
});
