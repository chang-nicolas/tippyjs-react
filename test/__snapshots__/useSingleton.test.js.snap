// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The useSingleton hook props.plugins 1`] = `
Array [
  Object {
    "defaultValue": false,
    "fn": [Function],
    "name": "followCursor",
  },
  Object {
    "defaultValue": false,
    "fn": [Function],
    "name": "inlinePositioning",
  },
  Object {
    "defaultValue": false,
    "fn": [Function],
    "name": "sticky",
  },
  Object {
    "fn": [Function],
  },
  Object {
    "fn": [Function],
  },
]
`;
